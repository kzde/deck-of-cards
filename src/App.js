import React, { Fragment } from 'react';
import Header from '../src/components/Header/Header';
import CardContainer from '../src/components/CardContainer/CardContainer';

const App = () => {
    return (
        <Fragment>
            <Header />
            <CardContainer />
        </Fragment>
    )
}

export default App;