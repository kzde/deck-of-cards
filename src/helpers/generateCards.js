const generateCards = () => {
    const cards = [];
    const groups = ['heart', 'spade', 'club', 'diamond'];
    const cardValues = ['2', '3', '4', '5', '6', '7', '8', '9', '10', 'Ace', 'Jack', 'Queen', 'King'];
    groups.forEach(group => {
        cardValues.forEach(value => {
            cards.push({
                group,
                value
            })
        })
    })
    return cards;
}

export default generateCards;
