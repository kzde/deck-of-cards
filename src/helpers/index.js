const getRandomIndex = v => Math.floor(Math.random() * Math.floor(v))

export const shuffle = array => {
    const arrLength = array.length;
    for (let i = arrLength - 1; i > 0; i--) {
        let randomIndex = getRandomIndex(i + 1);
        let tmp = array[i];
        array[i] = array[randomIndex];
        array[randomIndex] = tmp;
    }
    return array;
}

export const generateCards = () => {
    const cards = [];
    const groups = ['heart', 'spade', 'club', 'diamond'];
    const cardValues = ['Ace', '2', '3', '4', '5', '6', '7', '8', '9', '10', 'Jack', 'Queen', 'King'];
    groups.forEach(group => {
        cardValues.forEach(value => {
            cards.push({
                group,
                value
            })
        })
    })
    return cards;
}

export const dealOneCard = cards => {
    if (cards.length > 1) {
        const index = getRandomIndex(cards.length);
        const card = cards[index];
        cards.splice(index, 1);
        return { cards, card };
    } else if (cards.length === 1) {
        return { cards: [], card: cards[0] }
    }
    return { cards: [], card: {} };
}


