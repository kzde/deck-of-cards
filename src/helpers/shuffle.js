const shuffle = (array) => {
    const arrLength = array.length;
    for(let i = arrLength - 1 ; i > 0; i--){
        let randomIndex = getRandomIndex(i+1);
        let tmp = array[i];
        array[i] = array[randomIndex];
        array[randomIndex] = tmp;
    }
    return array;
}

const getRandomIndex = v => Math.floor(Math.random() * Math.floor(v))

export default shuffle;