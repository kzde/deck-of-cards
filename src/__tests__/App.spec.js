import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';
import App from '../App';
import * as helpers from '../helpers/index';
jest.mock('../components/Header/Header', () => 'Header');
jest.mock('../components/CardContainer/CardContainer', () => 'CardContainer');


describe('App', () => {

    it('should renders correctly', () => {
        const tree = renderer
            .create(<App />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    })

    describe('handleShuffle', () => {
        it('should call shuffle function and set card state to an empty object dealOneCard to be false', () => {
            const shuffleSpy = jest.fn();
            helpers.shuffle = shuffleSpy;
            const wrapper = shallow(<App />);
            wrapper.setState({
                card: { goupe: 'club', value: '2' },
                dealOneCard: true,
                shuffleForDealOneCard: true
            })
            wrapper.instance().handleShuffle();
            expect(shuffleSpy).toHaveBeenCalled();
            expect(wrapper.state('card')).toEqual({});
            expect(wrapper.state('dealOneCard')).toBeFalsy();
            expect(wrapper.state('shuffleForDealOneCard')).toBeFalsy();
        });

        it('should call generateCards function if there is no cards in the cards state', () => {
            const shuffleSpy = jest.fn();
            const generateCardsSpy = jest.fn().mockReturnValue([]);
            helpers.shuffle = shuffleSpy;
            helpers.generateCards = generateCardsSpy;
            const wrapper = shallow(<App />);
            wrapper.instance().handleShuffle();
            expect(generateCardsSpy).toHaveBeenCalled();
        })
    })

    describe('handleDealOneCard', () => {
        it('should call shuffle if the shuffleForDealOneCard is false', () => {
            const shuffleSpy = jest.fn();
            helpers.shuffle = shuffleSpy;
            const wrapper = shallow(<App />);
            wrapper.setState({
                shuffleForDealOneCard: false
            })
            wrapper.instance().handleDealOneCard();
            expect(shuffleSpy).toHaveBeenCalled();
        })

        it('should call dealOneCard function', () => {
            const dealOneCardSpy = jest.fn().mockReturnValue({
                card: { group: 'heart' },
                cards: [{ group: 'heart' }]
            });
            helpers.dealOneCard = dealOneCardSpy;
            const wrapper = shallow(<App />);
            wrapper.instance().handleDealOneCard();
            expect(dealOneCardSpy).toHaveBeenCalled();
        })

        it('should set state with the dealOneCard return value', () => {
            const dealOneCardSpy = jest.fn().mockReturnValue({
                card: { group: 'heart' },
                cards: [{ group: 'heart' }]
            });
            helpers.dealOneCard = dealOneCardSpy;
            const wrapper = shallow(<App />);
            wrapper.instance().handleDealOneCard();
            expect(wrapper.state('card')).toEqual({ group: 'heart' });
            expect(wrapper.state('cards')).toEqual([{ group: 'heart' }]);
        })
    })
})