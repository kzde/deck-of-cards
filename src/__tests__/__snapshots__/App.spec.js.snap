// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should renders correctly 1`] = `
Array [
  <Header
    handleDealOneCard={[Function]}
    handleShuffle={[Function]}
  />,
  <CardContainer
    card={Object {}}
    cards={
      Array [
        Object {
          "group": "heart",
          "value": "Ace",
        },
        Object {
          "group": "heart",
          "value": "2",
        },
        Object {
          "group": "heart",
          "value": "3",
        },
        Object {
          "group": "heart",
          "value": "4",
        },
        Object {
          "group": "heart",
          "value": "5",
        },
        Object {
          "group": "heart",
          "value": "6",
        },
        Object {
          "group": "heart",
          "value": "7",
        },
        Object {
          "group": "heart",
          "value": "8",
        },
        Object {
          "group": "heart",
          "value": "9",
        },
        Object {
          "group": "heart",
          "value": "10",
        },
        Object {
          "group": "heart",
          "value": "Jack",
        },
        Object {
          "group": "heart",
          "value": "Queen",
        },
        Object {
          "group": "heart",
          "value": "King",
        },
        Object {
          "group": "spade",
          "value": "Ace",
        },
        Object {
          "group": "spade",
          "value": "2",
        },
        Object {
          "group": "spade",
          "value": "3",
        },
        Object {
          "group": "spade",
          "value": "4",
        },
        Object {
          "group": "spade",
          "value": "5",
        },
        Object {
          "group": "spade",
          "value": "6",
        },
        Object {
          "group": "spade",
          "value": "7",
        },
        Object {
          "group": "spade",
          "value": "8",
        },
        Object {
          "group": "spade",
          "value": "9",
        },
        Object {
          "group": "spade",
          "value": "10",
        },
        Object {
          "group": "spade",
          "value": "Jack",
        },
        Object {
          "group": "spade",
          "value": "Queen",
        },
        Object {
          "group": "spade",
          "value": "King",
        },
        Object {
          "group": "club",
          "value": "Ace",
        },
        Object {
          "group": "club",
          "value": "2",
        },
        Object {
          "group": "club",
          "value": "3",
        },
        Object {
          "group": "club",
          "value": "4",
        },
        Object {
          "group": "club",
          "value": "5",
        },
        Object {
          "group": "club",
          "value": "6",
        },
        Object {
          "group": "club",
          "value": "7",
        },
        Object {
          "group": "club",
          "value": "8",
        },
        Object {
          "group": "club",
          "value": "9",
        },
        Object {
          "group": "club",
          "value": "10",
        },
        Object {
          "group": "club",
          "value": "Jack",
        },
        Object {
          "group": "club",
          "value": "Queen",
        },
        Object {
          "group": "club",
          "value": "King",
        },
        Object {
          "group": "diamond",
          "value": "Ace",
        },
        Object {
          "group": "diamond",
          "value": "2",
        },
        Object {
          "group": "diamond",
          "value": "3",
        },
        Object {
          "group": "diamond",
          "value": "4",
        },
        Object {
          "group": "diamond",
          "value": "5",
        },
        Object {
          "group": "diamond",
          "value": "6",
        },
        Object {
          "group": "diamond",
          "value": "7",
        },
        Object {
          "group": "diamond",
          "value": "8",
        },
        Object {
          "group": "diamond",
          "value": "9",
        },
        Object {
          "group": "diamond",
          "value": "10",
        },
        Object {
          "group": "diamond",
          "value": "Jack",
        },
        Object {
          "group": "diamond",
          "value": "Queen",
        },
        Object {
          "group": "diamond",
          "value": "King",
        },
      ]
    }
    dealOneCard={false}
  />,
]
`;
